### PR TITLE
Allow FlexVolume PV secret namespaces

### PR DESCRIPTION
**Pull request by liggitt**
Completes the secret namespace PV refactor, so all PV volume sources that specify secrets can reference them outside the PVC namespace.

Finished the secret-related aspect of https://github.com/kubernetes/kubernetes/issues/32131

```release-note
PersistentVolume flexVolume sources can now reference secrets in a namespace other than the PersistentVolumeClaim's namespace.
```

### CODE REVIEW

**Comment by liggitt** [comment]
cc @kubernetes/sig-storage-api-reviews @kubernetes/sig-storage-pr-reviews

**Comment by rootfs** [comment]
cc @chakri-nelluri

File: pkg/printers/internalversion/describe.go [hunk 1]
```diff
@@ -1076,6 +1076,16 @@ func printAzureFilePersistentVolumeSource(azureFile *api.AzureFilePersistentVolu
 		azureFile.SecretName, ns, azureFile.ShareName, azureFile.ReadOnly)
 }
 
+func printFlexPersistentVolumeSource(flex *api.FlexPersistentVolumeSource, w PrefixWriter) {
```
Comment (rootfs) [inline pkg/printers/internalversion/describe.go]: interesting, there is never a unit test for flex

**Comment by saad-ali** [comment]
CC @verult

File: pkg/volume/flexvolume/util.go [hunk 2]
```diff
@@ -52,15 +55,68 @@ func addSecretsToOptions(options map[string]string, spec *volume.Spec, namespace
 	return nil
 }
 
-func getVolumeSource(spec *volume.Spec) (volumeSource *api.FlexVolumeSource, readOnly bool) {
+var notFlexVolume = fmt.Errorf("not a flex volume")
+
+func getDriver(spec *volume.Spec) (string, error) {
```
Comment (verult) [inline pkg/volume/flexvolume/util.go]: Is this the driver name or driver path? The method name could be more clear

File: pkg/volume/flexvolume/util.go [hunk 3]
```diff
@@ -52,15 +55,68 @@ func addSecretsToOptions(options map[string]string, spec *volume.Spec, namespace
...
+	if spec.Volume != nil && spec.Volume.FlexVolume != nil {
+		return spec.Volume.FlexVolume.FSType, nil
+	}
+	if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.FlexVolume != nil {
+		return spec.PersistentVolume.Spec.FlexVolume.FSType, nil
+	}
+	return "", notFlexVolume
+}
+
+func getSecretNameAndNamespace(spec *volume.Spec, podNamespace string) (string, string, error) {
```
Comment (verult) [inline pkg/volume/flexvolume/util.go]: Consider using named return values

File: pkg/apis/core/types.go [hunk 4]
```diff
@@ -867,6 +867,32 @@ type FCVolumeSource struct {
...
+	// sensitive information to pass to the plugin scripts. This may be
+	// empty if no secret object is specified. If the secret object
+	// contains more than one secret, all secrets are passed to the plugin
+	// scripts.
+	// +optional
+	SecretRef *SecretReference
+	// Optional: Defaults to false (read/write). ReadOnly here will force
+	// the ReadOnly setting in VolumeMounts.
+	// +optional
+	ReadOnly bool
```
Comment (verult) [inline pkg/apis/core/types.go]: This field is never read

File: pkg/printers/internalversion/describe.go (hunk 1 above)
Comment (verult) [reply pkg/printers/internalversion/describe.go]: and secrets injection is not tested anywhere

File: pkg/apis/core/types.go (hunk 4 above)
Comment (liggitt) [reply pkg/apis/core/types.go]: this is an exact replica of the existing struct fields to start to make sure wire serialization is precisely compatible

File: pkg/printers/internalversion/describe.go [hunk 5]
```diff
@@ -1076,6 +1076,16 @@ func printAzureFilePersistentVolumeSource(azureFile *api.AzureFilePersistentVolu
...
+	w.Write(LEVEL_2, "Type:\tFlexVolume (a generic volume resource that is provisioned/attached using an exec based plugin)\n"+
+		"    Driver:\t%v\n"+
+		"    FSType:\t%v\n"+
+		"    SecretRef:\t%v\n"+
+		"    ReadOnly:\t%v\n",
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/printers/internalversion/describe.go]: ?

File: pkg/volume/flexvolume/util.go (hunk 2 above)
Comment (liggitt) [reply pkg/volume/flexvolume/util.go]: it's the driver field

File: pkg/printers/internalversion/describe.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (verult) [reply pkg/printers/internalversion/describe.go]: Did you mean to have a '+' at the end instead of ','?

[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/printers/internalversion/describe.go]: this was a copy/paste of the existing `printFlexVolumeSource` function with no changes other than the arg type

[the remark this answers is not in the document]
Comment (chakri-nelluri) [reply pkg/printers/internalversion/describe.go]: Nice..

**Review by chakri-nelluri — APPROVED** [review]

**Comment by chakri-nelluri** [comment]
Thanks @liggitt /LGTM /approve

File: pkg/apis/core/types.go [hunk 6]
```diff
@@ -354,7 +354,7 @@ type PersistentVolumeSource struct {
 	// FlexVolume represents a generic volume resource that is
 	// provisioned/attached using an exec based plugin. This is an alpha feature and may change in future.
 	// +optional
-	FlexVolume *FlexVolumeSource
+	FlexVolume *FlexPersistentVolumeSource
```
Comment (chakri-nelluri) [inline pkg/apis/core/types.go]: *Nit - do we need to repeat the Persistent in FlexPersistentVolumeSource? I like FlexVolumeSource

**Comment by liggitt** [comment]
rebased, retagging

**Comment by liggitt** [comment]
regenerated

File: pkg/printers/internalversion/describe.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (php-coder) [reply pkg/printers/internalversion/describe.go]: BTW it's already fixed: https://github.com/kubernetes/kubernetes/pull/57202

[the remark this answers is not in the document]
Comment (php-coder) [reply pkg/printers/internalversion/describe.go]: I've submitted another one to fix this particular case: https://github.com/kubernetes/kubernetes/pull/57399

**Pull request merged**